fusedrug/data/protein/sequence/official_pdb_fasta.py

```diff
@@ -1,38 +1,141 @@
+def get_fasta_from_rcsb(
+    pdb_id: str,
+    author_chain_id: bool = True,
+    verbose: bool = False,
+    return_duplicates_if_found: bool = True,
+) -> Dict:  # TODO: consider adding caching
     """
     Given some pdb_id, (like "7vux"), we will retrieve its fasta file from rcsb database and return it as a dict {chain: sequence}.
+    This version handles symmetry, so if there are multiple identical chains in the same complex, it will return them as separate keys
     """
     fasta_data = (
         urlopen(f"https://www.rcsb.org/fasta/entry/{pdb_id.upper()}")
         .read()
         .decode("utf-8")
     )
-    fasta_file_handle = StringIO(fasta_data)
-    chains_full_seq = SeqIO.to_dict(
-        SeqIO.parse(fasta_file_handle, "fasta"),
-        key_function=lambda rec: _description_to_author_chain_id(rec.description),
-    )
-    chains_full_seq = {k: str(d.seq) for (k, d) in chains_full_seq.items()}
-    return chains_full_seq
+
+    ans = {}
+
+    io_data = StringIO(fasta_data)
+    for record in SeqIO.parse(io_data, "fasta"):
+        print(f"{record.description=}")
+        chains = get_chain_ids(
+            description=record.description, author_chain_id=author_chain_id
+        )
+        print(f"{chains=}")
+
+        for chain in chains:
+            ans[chain] = str(record.seq)
+            if not return_duplicates_if_found:
+                break
+
+    return ans
+
+
+def get_chain_ids(*, description: str, author_chain_id: bool) -> List[str]:
+    """
+    author_chain_id: if True will return the author chain id, if False will return the pdb_chain_id
+    TODO: add support for chain ids that have more than one character
+    """
+    loc1 = description.find(" ")
+    assert loc1 >= 0
+
+    description = description[loc1 + 1 :]
+
+    loc2 = description.find("|")
+    assert loc2 >= 0
+    description = description[:loc2].strip()
+
+    ans = []
+
+    for part in description.split(","):
+        part = part.strip()
+
+        curr_pdb_chain_id = part[0]
+        curr_author_chain_id = curr_pdb_chain_id
+        if author_chain_id:
+            if " " in part:
+                curr_author_chain_id = part[part.find(" ") + 1]
+
+        if author_chain_id:
+            ans.append(curr_author_chain_id)
+        else:
+            ans.append(curr_pdb_chain_id)
+
+    return ans
 
 
-def _description_to_author_chain_id(description: str) -> str:
-    loc = description.find(" ")
-    assert loc >= 0
-    description = description[loc + 1 :]
-    loc = description.find(",")
-    if loc >= 0:
-        description = description[:loc]
+# examples for reference:
 
-    token = "auth "
-    loc = description.find(token)
-    if loc >= 0:
-        return description[loc + len(token)]
+# >6N25_1|Chains A, B, C, D, E|Bestrophin homolog|Gallus gallus (9031)
+#
 
-    return description[0]
+# seq.description='9DH2_1|Chains A, C[auth G], E[auth K], H[auth P]|Fab heavy chain|Homo sapiens (9606)'
+# chains=['A', 'C', 'E', 'H']
```

Comment:
these examples show the chains in pdb assigned or author assigned form?

Reply:
the examples show whatever is in the original FASTA file, some only have pdb assigned and some have both pdb assigned and author assigned